File: src/main/java/org/example/kalkulationsprogramm/domain/Textbaustein.java
package org.example.kalkulationsprogramm.domain;

import jakarta.persistence.*;
import java.time.OffsetDateTime;
import java.util.LinkedHashSet;
import java.util.Set;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@Entity
@Table(name = "textbaustein")
public class Textbaustein {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 150)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(name = "typ", nullable = false, length = 40)
    private TextbausteinTyp typ = TextbausteinTyp.FREITEXT;

    @Column(name = "beschreibung", length = 500)
    private String beschreibung;

    @Column(name = "html", columnDefinition = "longtext")
    private String html;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "textbaustein_placeholder", joinColumns = @JoinColumn(name = "textbaustein_id"))
    @Column(name = "placeholder", length = 120)
    private Set<String> placeholders = new LinkedHashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @Enumerated(EnumType.STRING)
    @CollectionTable(name = "textbaustein_dokumenttyp_enum", joinColumns = @JoinColumn(name = "textbaustein_id"))
    @Column(name = "dokumenttyp", length = 30, columnDefinition = "varchar(30)")
    private Set<Dokumenttyp> dokumenttypen = new LinkedHashSet<>();

    @Column(name = "sort_order")
    private Integer sortOrder;

    @Column(name = "created_at")
    private OffsetDateTime createdAt;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;

    @PrePersist
    public void onCreate() {
        createdAt = OffsetDateTime.now();
        updatedAt = createdAt;
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/Textbaustein.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.*
import java.time.OffsetDateTime

@Entity
@Table(name = "textbaustein")
open class Textbaustein {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @Column(nullable = false, length = 150)
    open var name: String? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "typ", nullable = false, length = 40)
    open var typ: TextbausteinTyp? = TextbausteinTyp.FREITEXT

    @Column(name = "beschreibung", length = 500)
    open var beschreibung: String? = null

    @Column(name = "html", columnDefinition = "longtext")
    open var html: String? = null

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "textbaustein_placeholder")
    @Column(name = "placeholder", length = 120)
    open var placeholders: MutableSet<String> = linkedSetOf()

    @ElementCollection(fetch = FetchType.EAGER)
    @Enumerated(EnumType.STRING)
    @CollectionTable(name = "textbaustein_dokumenttyp_enum")
    @Column(name = "dokumenttyp", length = 30, columnDefinition = "varchar(30)")
    open var dokumenttypen: MutableSet<Dokumenttyp> = linkedSetOf()

    @Column(name = "sort_order")
    open var sortOrder: Int? = null

    @Column(name = "created_at")
    open var createdAt: OffsetDateTime? = null

    @Column(name = "updated_at")
    open var updatedAt: OffsetDateTime? = null

}
